File: example/1d-linear-convection/weno3/python-v1/01c/initial_condition.py
# initial_condition.py

"""
初始条件模块 (已集成注册系统)
使用装饰器 @register_component 自动注册
"""

import numpy as np
from abc import ABC, abstractmethod
from cfd_registry import ComponentRegistry, register_component

# ---------------------- 1. 初始条件抽象基类 ----------------------
class InitialCondition(ABC):
    """初始条件基类"""
    def __init__(self, config):
        self.config = config

    @abstractmethod
    def apply(self, solution):
        """将初始条件应用到 solution 的内部区域"""
        pass

    @abstractmethod
    def evaluate_at(self, x):
        """纯数学函数：给定 x，返回 u(x)，不涉及网格或边界"""
        pass

    def exact_solution(self, cfd):
        """
        默认解析解：线性对流 u(x, t) = u0(x - c * t)
        子类可重写以支持更复杂物理
        """
        x = cfd.domain.mesh.xcc
        t = cfd.config.final_time
        c = cfd.config.wave_speed
        L = cfd.domain.mesh.L
        # 周期平移：确保在 [x0, x0 + L) 内
        x_shifted = (x - c * t + L) % L        
        return self.evaluate_at(x_shifted)

    def _apply_to_interior(self, solution, values):
        domain = solution.domain
        for i in range(domain.ist, domain.ied):
            j = i - domain.ist
            solution.u[i] = values[j]


# ---------------------- 2. 具体初始条件实现（使用装饰器注册） ----------------------

@register_component('initial_condition', 'step')
class StepFunctionIC(InitialCondition):
    def evaluate_at(self, x):
        u0 = np.ones_like(x)
        mask = (x >= 0.5) & (x <= 1.0)
        u0[mask] = 2.0
        return u0

    def apply(self, solution):
        x = solution.domain.mesh.xcc
        u0 = self.evaluate_at(x)
        self._apply_to_interior(solution, u0)

@register_component('initial_condition', 'sin')
class SineWaveIC(InitialCondition):
    def evaluate_at(self, x):
        L = getattr(self.config, "domain_length", 2.0)
        return np.sin(2 * np.pi * x / L)

    def apply(self, solution):
        x = solution.domain.mesh.xcc
        u0 = self.evaluate_at(x)
        self._apply_to_interior(solution, u0)

@register_component('initial_condition', 'gaussian')
class GaussianPulseIC(InitialCondition):
    def evaluate_at(self, x):
        center = getattr(self.config, "pulse_center", 0.5)
        width = getattr(self.config, "pulse_width", 0.1)
        return np.exp(-((x - center) / width) ** 2)

    def apply(self, solution):
        x = solution.domain.mesh.xcc
        u0 = self.evaluate_at(x)
        self._apply_to_interior(solution, u0)

class InitialConditionFactory:
    """初始条件工厂"""
    
    @classmethod
    def create(cls, config) -> 'InitialCondition':
        """
        创建初始条件实例
        
        Args:
            ic_type: 初始条件类型（如'step', 'sin'等）
            config: 配置对象
        
        Returns:
            初始条件实例
        """
        from factories.base_factory import BaseFactory
        return BaseFactory.create_component('initial_condition', config.ic_type, config)
        
    @classmethod
    def get_exact_solution(cls, cfd):
        return cls.create(cfd.config).exact_solution(cfd)  # 一行！        
        

File: example/1d-linear-convection/weno3/julia/01d/julia/initial_condition.jl
# julia/initial_condition.jl
"""
初始条件模块（Julia 版）
目标：与 Python initial_condition.py 行为完全一致
"""

# ---------------------- 辅助函数 ----------------------
"""
模拟 Python 的 getattr(obj, name, default)
"""
function getfield_safe(obj, name::Symbol, default)
    return isdefined(obj, name) ? getfield(obj, name) : default
end

# ---------------------- 抽象接口（Julia 无继承，用函数约定） ----------------------
# 所有初始条件必须实现：
#   evaluate_at(ic, x::AbstractVector{Float64}) -> Vector{Float64}
#   apply(ic, solution)

# ---------------------- StepFunctionIC ----------------------
struct StepFunctionIC
    config::Any
end

function evaluate_at(ic::StepFunctionIC, x::AbstractVector{Float64})
    u0 = ones(Float64, length(x))
    for i in eachindex(x)
        if 0.5 <= x[i] <= 1.0
            u0[i] = 2.0
        end
    end
    return u0
end

function apply(ic::StepFunctionIC, solution)
    x = solution.domain.mesh.xcc
    u0 = evaluate_at(ic, x)
    _apply_to_interior(solution, u0)
end

# ---------------------- SineWaveIC ----------------------
struct SineWaveIC
    config::Any
end

function evaluate_at(ic::SineWaveIC, x::AbstractVector{Float64})
    L = getfield_safe(ic.config, :domain_length, 2.0)
    return sin.(2π * x / L)
end

function apply(ic::SineWaveIC, solution)
    x = solution.domain.mesh.xcc
    u0 = evaluate_at(ic, x)
    _apply_to_interior(solution, u0)
end

# ---------------------- GaussianPulseIC ----------------------
struct GaussianPulseIC
    config::Any
end

function evaluate_at(ic::GaussianPulseIC, x::AbstractVector{Float64})
    center = getfield_safe(ic.config, :pulse_center, 0.5)
    width = getfield_safe(ic.config, :pulse_width, 0.1)
    return exp.(-((x .- center) ./ width).^2)
end

function apply(ic::GaussianPulseIC, solution)
    x = solution.domain.mesh.xcc
    u0 = evaluate_at(ic, x)
    _apply_to_interior(solution, u0)
end

# ---------------------- 公共辅助函数 ----------------------
"""
将初始场 values 应用到 solution.u 的物理区域（含 ghost 的数组）
"""
function _apply_to_interior(solution, values)
    domain = solution.domain
    # Python: for i in range(domain.ist, domain.ied)
    # Julia:  for i in domain.ist:domain.ied-1 （因为 ied 是结束索引，不包含）
    for i in domain.ist:(domain.ied - 1)
        j = i - domain.ist + 1  # values 是 1-based，i - ist 是 0-based → +1
        solution.u[i] = values[j]
    end
end